tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc

```diff
@@ -1090,7 +1130,8 @@ std::vector<int> LayoutAgnosticOpTransposer::GetVariadic4DFaninPorts(
     const auto& regular_fanin = node.GetRegularFanin(i);
     auto* regular_fanin_node = regular_fanin.node_view();
     int regular_fanin_port = regular_fanin.index();
-    if (IsFanoutPortRankN(*regular_fanin_node, regular_fanin_port, 4) &&
+    if ((IsFanoutPortRankN(*regular_fanin_node, regular_fanin_port, 4) ||
+         IsFanoutPortRankN(*regular_fanin_node, regular_fanin_port, 5)) &&
         ((IsAfterDstToSrcTransform(context, *regular_fanin_node) &&
```

Comment:
Does this support a mixture of say 4D and 5D tensors? I assume after changing the layout, only 5D tensors will be detected due to IsAfterDstToSrcTransform? And can a test case be added for this (ShapeN and IdentityN taking in ranked 4 and ranked 5 tensors)?

Reply:
Can ShapeN take in tensors with different dims? I tried to give shape_n two different tensors but got `tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimensions must be equal, but are 5 and 4 for '{{node Add_2}} = Add[T=DT_INT32](ShapeN, ShapeN:1)' with input shapes: [5], [4].`

Reply:
It looks like ShapeN requires same ranks for all arguments. I don't believe IdentityN has that restriction though.

Reply:
Done. Please check the new commit.